### PR TITLE
Unify restore across Windows and Linux

### PR DESCRIPTION
**Pull request by jaredpar**
This changes the Windows builds to restore in the same manner as Linux.  Helps ensure we keep the lab consistent between the runs.

The other positive upside of this change is shaves ~10 minutes of individual runs.  Roslyn.sln is a *huge* solution, has a rather large MSBuild include pattern and as a result puts a lot of pressure on restore.  As such the set of restores we do typically takes ~10 minutes.

Long term we'll be working with NuGet to get this time down a bit.  At the moment though we're a bit of an odd solution here with our size and time requirements.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Looks good to me.  But I should probably defer this to someone who knows better how the infrastructure pieces work.

File: build/scripts/restore.ps1 [hunk 1]
```diff
@@ -0,0 +1,13 @@
+param ([string]$nugetZipUrl = $(throw "Need an URL to the NuGet zip") )
+
+$destination = ${env:UserProfile}
+$outFilePath = [IO.Path]::ChangeExtension([IO.Path]::GetTempFileName(), "zip")
```
Comment (CyrusNajmabadi) [inline build/scripts/restore.ps1]: Do you even need to bother changing the extension?

File: build/scripts/restore.ps1 [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+write-host "Downloading $nugetZipUrl -> $outFilePath"
+$client = new-object System.Net.WebClient
+$client.DownloadFile($nugetZipUrl, $outFilePath)
+
+write-host "Extracting"
+Add-Type -assembly "System.IO.Compression.Filesystem"
+$archive = [IO.Compression.ZipFile]::OpenRead($outFilePath)
+rm -re (join-path $destination ".nuget\packages") -errorAction SilentlyContinue
```
Comment (CyrusNajmabadi) [inline build/scripts/restore.ps1]: SilentlyContinue?

File: build/scripts/restore.ps1 (hunk 1 above)
Comment (jaredpar) [inline build/scripts/restore.ps1]: It was necessary when I was using the shell to do the unzip.  Now i may be able to avoid it.

File: build/scripts/restore.ps1 (hunk 2 above)
Comment (jaredpar) [inline build/scripts/restore.ps1]: Powershell has two error modes:

1. Spew gobs of unreadable red text to the screen 
2. Utter silence

Comment (CyrusNajmabadi) [inline build/scripts/restore.ps1]: Wouldn't the former be better?  Or are you saying "it's ok and expected that we should not be able to remove this directory, and we want to keep on truckin gin that case?"

Comment (jaredpar) [inline build/scripts/restore.ps1]: It's the latter.  

About to submit another commit which changes tihs a bit.  Forgot we had to deal with cooperative restore scenarios.

File: cibuild.cmd [hunk 3]
```diff
@@ -11,19 +14,27 @@ if /I "%1" == "/debug" set BuildConfiguration=Debug&&shift&& goto :ParseArgument
...
 call "C:\Program Files (x86)\Microsoft Visual Studio 14.0\Common7\Tools\VsDevCmd.bat"
 
-REM Build the compiler so we can self host it for the full build
-nuget.exe restore -nocache -verbosity quiet %RoslynRoot%build/ToolsetPackages/project.json
-nuget.exe restore -nocache -verbosity quiet %RoslynRoot%build/Toolset.sln
+REM Restore the NuGet packages 
+if "%BuildRestore%" == "true" (
+    nuget.exe restore -nocache -verbosity quiet %RoslynRoot%build/ToolsetPackages/project.json
+    nuget.exe restore -nocache -verbosity quiet %RoslynRoot%build/Toolset.sln
+    nuget.exe restore -nocache %RoslynRoot%build\ToolsetPackages\project.json
```
Comment (Pilchie) [inline cibuild.cmd]: You restore this twice, both with and without -verbosity quiet.

Comment (jaredpar) [inline cibuild.cmd]: Ha.  I just copied what was already there.  Didn't think to look at it.  But yes looks like a duplicate.

File: build/scripts/restore.ps1 [hunk 4]
```diff
@@ -0,0 +1,26 @@
+param ([string]$nugetZipUrl = $(throw "Need an URL to the NuGet zip") )
+
+$destination = ${env:UserProfile}
+$outFilePath = [IO.Path]::ChangeExtension([IO.Path]::GetTempFileName(), "zip")
+write-host "Downloading $nugetZipUrl -> $outFilePath"
+$client = new-object System.Net.WebClient
+$client.DownloadFile($nugetZipUrl, $outFilePath)
```
Comment (Pilchie) [inline build/scripts/restore.ps1]: Shouldn't you trap this too in case of network issues?

File: build/scripts/restore.ps1 [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+# It's possible for restore to run in parallel on the test machines.  As such
+# we need to restore only new files to handle simultaneous restore scenarios.
```
Comment (Pilchie) [inline build/scripts/restore.ps1]: I don't really like this at all.  I thought @agocke added a restore semaphore to the max/linux builds.  Why not do the same here?

Comment (jaredpar) [inline build/scripts/restore.ps1]: @Pilchie we removed the restore semaphore because it's an unworkable solution.  To be correct the semaphore would need to wrap the entire build step.  

IMHO: doing parallel runs on a machine is a mistake.  So long as we have global assets, like NuGet packages, we're always going to be hurting ourselves.  But right now that's how it is.

Comment (tannergooding) [inline build/scripts/restore.ps1]: @jaredpar, @Pilchie. A normal nuget restore (in v3 where it goes to the %UserProfile%\.nuget folder) there is no issue restoring from multiple builds simultaneously, as a restore does not delete or overwrite what already exists.

Provided this script can do the same (unzip without touching an existing file), then their should be no issue with concurrent restore.

Overall, parallel runs on the machine is the best option, because the machines can handle the runs and it is a more manageable solution to have 10 machines which can do 2 runs each, then it is to have 20 machines.
